File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int longest = 0;
        char[] sArray = s.toCharArray();
        for(int left = 0, right = 0;right < sArray.length;right++){
            char current = sArray[right];
            while(set.contains(current) && left < right){
                set.remove(sArray[left++]);
            }
            longest = Math.max(longest, ( right - left ) + 1);
            set.add(current);
        }
        return longest;
    }
}

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        
        var map = HashMap<Char,Int>()
        var left = 0
        var longest = 0
        
        var sArray = s.toCharArray()
        
        for(i in 0 until sArray.size){
            val currentChar = sArray[i]
            
            
            map[currentChar]?.let{ currentIndex ->
                left = Math.max( left, currentIndex )
            }
            
            longest = Math.max( longest, i - left + 1 )
            map[currentChar] = i + 1
            
            
        }
        
        return longest
    }
}